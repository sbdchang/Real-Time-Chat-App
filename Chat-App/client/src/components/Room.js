import React, { useState, useEffect } from 'react';
import Video from 'twilio-video';
import Participant from './Participant';

const Room = ({ roomName, token, handleLogout }) => {
  const [room, setRoom] = useState(null);
  const [participants, setParticipants] = useState([]);

  useEffect(() => {
    const participantConnected = participant => {
      setParticipants(prevParticipants => [...prevParticipants, participant]);
    };

    const participantDisconnected = participant => {
      setParticipants(prevParticipants =>
        prevParticipants.filter(p => p !== participant)
      );
    };

    var minutesLabel = document.getElementById("minutes");
    var secondsLabel = document.getElementById("seconds");
    var totalSeconds = 0;
    setInterval(setTime, 1000);

    function setTime() {
      ++totalSeconds;
      secondsLabel.innerHTML = pad(totalSeconds % 60);
      minutesLabel.innerHTML = pad(parseInt(totalSeconds / 60));
    }

    function pad(val) {
      var valString = val + "";
      if (valString.length < 2) {
        return "0" + valString;
      } else {
        return valString;
      }
    }

    Video.connect(token, {
      name: roomName
    }).then(room => {
      setRoom(room);
      room.on('participantConnected', participantConnected);
      room.on('participantDisconnected', participantDisconnected);
      room.participants.forEach(participantConnected);
    });

    return () => {
      setRoom(currentRoom => {
        if (currentRoom && currentRoom.localParticipant.state === 'connected') {
          currentRoom.localParticipant.tracks.forEach(function(trackPublication) {
            trackPublication.track.stop();
          });
          currentRoom.disconnect();
          return null;
        } else {
          return currentRoom;
        }
      });
    };
  }, [roomName, token]);

  const remoteParticipants = participants.map(participant => (
    <Participant key={participant.sid} participant={participant} />
  ));

  return (
    <div className="room">
      <h2>Room: {roomName}</h2>
      <button onClick={handleLogout}>End call</button> <br /> <br />
      <div style={{ color: 'white'}}>Time elapsed <label id="minutes">00</label>:<label id="seconds">00</label></div>

      <div className="local-participant">
        {room ? (
          <Participant
            key={room.localParticipant.sid}
            participant={room.localParticipant}
          />
        ) : (
          ''
        )}
      </div>
      <h3>Remote Participants</h3>
      <div className="remote-participants">{remoteParticipants}</div>
    </div>
  );
};

export default Room;
