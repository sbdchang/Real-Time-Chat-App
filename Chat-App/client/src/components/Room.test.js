import React from 'react';
import ReactDOM from 'react-dom';
import { render, screen } from '@testing-library/react';
import Room from './Room';
import renderer from 'react-test-renderer';



it('Testing here', () => {
  expect(1).toBe(1);
});


// it('Room snapshot   ', () => {
//   const tree = renderer.create(<Room />).toJSON();
//   expect(tree).toMatchSnapshot();
// });



// it('renders without crashing', () => {
//   const div = document.createElement('div');
//   ReactDOM.render(<Room roomName={"roomName"} token={"token"} handleLogout={"handleLogout"} />, div);
// });
