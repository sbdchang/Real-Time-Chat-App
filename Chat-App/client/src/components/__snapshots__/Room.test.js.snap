// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Room snapshot    1`] = `
<div
  className="room"
>
  <h2>
    Room: 
  </h2>
  <button>
    End call
  </button>
   
  <br />
   
  <br />
  <div
    style={
      Object {
        "color": "white",
      }
    }
  >
    Time elapsed 
    <label
      id="minutes"
    >
      00
    </label>
    :
    <label
      id="seconds"
    >
      00
    </label>
  </div>
  <div
    className="local-participant"
  >
    
  </div>
  <h3>
    Remote Participants
  </h3>
  <div
    className="remote-participants"
  />
</div>
`;
